macros/src/main/scala/mainecoon/autoFlatMap.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2017 Kailuo Wang
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mainecoon
+
+import mainecoon.Util._
+
+import scala.annotation.{StaticAnnotation, compileTimeOnly}
+import scala.collection.immutable.Seq
+import scala.meta._
+
+/**
+ * auto generates an instance of `cats.FlatMap`
+ */
+@compileTimeOnly("Cannot expand @autoMonad")
+class autoFlatMap extends StaticAnnotation {
+  inline def apply(defn: Any): Any = meta {
+    enrichAlgebra(defn, higherKinded = false)(autoFlatMap.flatMapInst)
+  }
+}
+
+
+object autoFlatMap {
+  private[mainecoon] def flatMapInst(ad: AlgDefn): TypeDefinition = {
+    import ad._
+    import cls._
+
+
+    val mapMethods = templ.stats.toList.flatMap(_.collect {
```

Comment:
The duplication of this mapMethods code between `autoFlatMap` and `autoFunctor` is a pity. Is there a straightforward way to reduce it a little bit? (maybe a method that takes a `ad: AlgDefn` and the mapFunction name and delegatee name?) what do you think? What do you think, at the very minimum we can leave a comment here suggesting that it's duplicated with `autoFunctor`.

Reply:
I created a common `mapMethods` function in autoFlatMap.